### ISSUE TITLE
MIR-borrowck: (unsound) fails to see errors in borrowck-vec-pattern-move-tail.rs

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#44834]
Many MIR-borrowck bugs have been categorized as either extra errors or poor diagnostics, but this is a case where MIR borrowck is simply failing to flag *the* main error in the code in question.

We see below that AST-borrowck is flagging the error while MIR-borrowck says nothing. (This is *atop* the mir-borrowck4 branch from https://github.com/rust-lang/rust/pull/44736 )

```
 % ./objdir-dbgopt/build/x86_64-unknown-linux-gnu/stage1/bin/rustc -Z emit-end-regions \
    -Z borrowck-mir ./src/test/compile-fail/borrowck/borrowck-vec-pattern-move-tail.rs
error[E0506]: cannot assign to `a[..]` because it is borrowed (Ast)
  --> ./src/test/compile-fail/borrowck/borrowck-vec-pattern-move-tail.rs:20:5
   |
16 |         [1, 2, ref tail..] => tail,
   |                -------- borrow of `a[..]` occurs here
...
20 |     a[2] = 0; //~ ERROR cannot assign to `a[..]` because it is borrowed
   |     ^^^^^^^^ assignment to borrowed `a[..]` occurs here

error: aborting due to previous error
```
